File: src/Interpreter.cpp
#include "Idot/Interpreter.h"

#include <cmath>
#include <memory>
#include <utility>

#include "Idot/Diagnostics.h"

namespace idot {
namespace {

template <class... Ts>
struct Overloaded : Ts... {
  using Ts::operator()...;
};

template <class... Ts>
Overloaded(Ts...) -> Overloaded<Ts...>;

}  // namespace

Interpreter::Interpreter() : environment_(std::make_shared<Environment>()) {}

void Interpreter::Execute(const std::vector<StmtPtr>& statements, std::ostream& output) {
  for (const auto& statement : statements) {
    Execute(*statement, output);
  }
}

void Interpreter::Execute(const Stmt& statement, std::ostream& output) {
  std::visit(
      Overloaded{
          [&](const BlockStmt& block) {
            ExecuteBlock(block.statements, std::make_shared<Environment>(environment_), output);
          },
          [&](const ExprStmt& expression) {
            (void)Evaluate(*expression.expression, output);
          },
          [&](const IfStmt& ifStatement) {
            const Value condition = Evaluate(*ifStatement.condition, output);
            if (IsTruthy(condition)) {
              Execute(*ifStatement.thenBranch, output);
            } else if (ifStatement.elseBranch) {
              Execute(*ifStatement.elseBranch, output);
            }
          },
          [&](const PrintStmt& printStatement) {
            const Value value = Evaluate(*printStatement.expression, output);
            output << value.ToString() << '\n';
          },
          [&](const VarStmt& variableStatement) {
            const Value value = Evaluate(*variableStatement.initializer, output);
            environment_->Define(variableStatement.name.lexeme, value);
          },
      },
      statement.node);
}

void Interpreter::ExecuteBlock(const std::vector<StmtPtr>& statements,
                               std::shared_ptr<Environment> environment, std::ostream& output) {
  auto previous = environment_;
  environment_ = std::move(environment);
  struct ScopeReset {
    std::shared_ptr<Environment>& slot;
    std::shared_ptr<Environment> previous;
    ~ScopeReset() { slot = std::move(previous); }
  } reset{environment_, previous};

  for (const auto& statement : statements) {
    Execute(*statement, output);
  }
}

Value Interpreter::Evaluate(const Expr& expression, std::ostream& output) {
  return std::visit(
      Overloaded{
          [&](const AssignExpr& assign) {
            const Value value = Evaluate(*assign.value, output);
            environment_->Assign(assign.name, value);
            return value;
          },
          [&](const BinaryExpr& binary) {
            const Value left = Evaluate(*binary.left, output);
            const Value right = Evaluate(*binary.right, output);

            switch (binary.op.type) {
              case TokenType::Plus:
                if (left.IsNumber() && right.IsNumber()) {
                  return Value(left.AsNumber() + right.AsNumber());
                }
                if (left.IsString() && right.IsString()) {
                  return Value(left.AsString() + right.AsString());
                }
                throw DiagnosticError(ErrorPhase::Runtime, binary.op.line, binary.op.column,
                                      "Operator '+' requires two numbers or two strings.");
              case TokenType::Minus:
                return Value(RequireNumber(left, binary.op, "left operand") -
                             RequireNumber(right, binary.op, "right operand"));
              case TokenType::Star:
                return Value(RequireNumber(left, binary.op, "left operand") *
                             RequireNumber(right, binary.op, "right operand"));
              case TokenType::Slash: {
                const double divisor = RequireNumber(right, binary.op, "right operand");
                if (divisor == 0.0) {
                  throw DiagnosticError(ErrorPhase::Runtime, binary.op.line, binary.op.column,
                                        "Division by zero.");
                }
                return Value(RequireNumber(left, binary.op, "left operand") / divisor);
              }
              case TokenType::Percent: {
                const double divisor = RequireNumber(right, binary.op, "right operand");
                if (divisor == 0.0) {
                  throw DiagnosticError(ErrorPhase::Runtime, binary.op.line, binary.op.column,
                                        "Modulo by zero.");
                }
                return Value(std::fmod(RequireNumber(left, binary.op, "left operand"), divisor));
              }
              case TokenType::Greater:
                return Value(RequireNumber(left, binary.op, "left operand") >
                             RequireNumber(right, binary.op, "right operand"));
              case TokenType::GreaterEqual:
                return Value(RequireNumber(left, binary.op, "left operand") >=
                             RequireNumber(right, binary.op, "right operand"));
              case TokenType::Less:
                return Value(RequireNumber(left, binary.op, "left operand") <
                             RequireNumber(right, binary.op, "right operand"));
              case TokenType::LessEqual:
                return Value(RequireNumber(left, binary.op, "left operand") <=
                             RequireNumber(right, binary.op, "right operand"));
              case TokenType::EqualEqual:
                return Value(left == right);
              case TokenType::BangEqual:
                return Value(left != right);
              default:
                break;
            }
            throw DiagnosticError(ErrorPhase::Runtime, binary.op.line, binary.op.column,
                                  "Unsupported binary operator.");
          },
          [&](const GroupingExpr& grouping) { return Evaluate(*grouping.expression, output); },
          [&](const LiteralExpr& literal) { return literal.value; },
          [&](const UnaryExpr& unary) {
            const Value right = Evaluate(*unary.right, output);
            switch (unary.op.type) {
              case TokenType::Bang:
                return Value(!IsTruthy(right));
              case TokenType::Minus:
                return Value(-RequireNumber(right, unary.op, "operand"));
              default:
                break;
            }
            throw DiagnosticError(ErrorPhase::Runtime, unary.op.line, unary.op.column,
                                  "Unsupported unary operator.");
          },
          [&](const VariableExpr& variable) { return environment_->Get(variable.name); },
      },
      expression.node);
}

bool Interpreter::IsTruthy(const Value& value) const {
  if (value.IsNil()) {
    return false;
  }
  if (value.IsBool()) {
    return value.AsBool();
  }
  return true;
}

double Interpreter::RequireNumber(const Value& value, const Token& token, const char* context) const {
  if (!value.IsNumber()) {
    throw DiagnosticError(ErrorPhase::Runtime, token.line, token.column,
                          std::string("Expected number for ") + context + ".");
  }
  return value.AsNumber();
}

}  // namespace idot

File: compiler/src/interpreter.rs
use std::cell::RefCell;
use std::io::Write;
use std::rc::Rc;

use crate::ast::{Expr, Stmt};
use crate::diagnostics::{DiagnosticError, ErrorPhase, Result};
use crate::environment::Environment;
use crate::token::{Token, TokenType};
use crate::value::Value;

pub struct Interpreter {
    environment: Rc<RefCell<Environment>>,
}

impl Interpreter {
    pub fn new() -> Self {
        Self {
            environment: Rc::new(RefCell::new(Environment::new(None))),
        }
    }

    pub fn execute(&mut self, statements: &[Stmt], output: &mut dyn Write) -> Result<()> {
        for statement in statements {
            self.execute_stmt(statement, output)?;
        }
        Ok(())
    }

    fn execute_stmt(&mut self, statement: &Stmt, output: &mut dyn Write) -> Result<()> {
        match statement {
            Stmt::Block(statements) => {
                let environment = Rc::new(RefCell::new(Environment::new(Some(
                    self.environment.clone(),
                ))));
                self.execute_block(statements, environment, output)
            }
            Stmt::Expr(expression) => {
                let _ = self.evaluate(expression)?;
                Ok(())
            }
            Stmt::If {
                condition,
                then_branch,
                else_branch,
            } => {
                if self.evaluate(condition)?.is_truthy() {
                    self.execute_stmt(then_branch, output)
                } else if let Some(branch) = else_branch {
                    self.execute_stmt(branch, output)
                } else {
                    Ok(())
                }
            }
            Stmt::Print(expression) => {
                let value = self.evaluate(expression)?;
                writeln!(output, "{}", value.to_idot_string())
                    .map_err(|error| Self::io_error("write failed", error))?;
                Ok(())
            }
            Stmt::Var { name, initializer } => {
                let value = self.evaluate(initializer)?;
                self.environment.borrow_mut().define(&name.lexeme, value);
                Ok(())
            }
        }
    }

    fn execute_block(
        &mut self,
        statements: &[Stmt],
        environment: Rc<RefCell<Environment>>,
        output: &mut dyn Write,
    ) -> Result<()> {
        let previous = self.environment.clone();
        self.environment = environment;

        let result = (|| {
            for statement in statements {
                self.execute_stmt(statement, output)?;
            }
            Ok(())
        })();

        self.environment = previous;
        result
    }

    fn evaluate(&mut self, expression: &Expr) -> Result<Value> {
        match expression {
            Expr::Assign { name, value } => {
                let assigned = self.evaluate(value)?;
                self.environment
                    .borrow_mut()
                    .assign(name, assigned.clone())?;
                Ok(assigned)
            }
            Expr::Binary { left, op, right } => {
                let left = self.evaluate(left)?;
                let right = self.evaluate(right)?;
                match op.kind {
                    TokenType::Plus => match (left, right) {
                        (Value::Number(lhs), Value::Number(rhs)) => Ok(Value::Number(lhs + rhs)),
                        (Value::String(lhs), Value::String(rhs)) => {
                            Ok(Value::String(format!("{lhs}{rhs}")))
                        }
                        _ => Err(Self::runtime_error(
                            op,
                            "Operator '+' requires two numbers or two strings.",
                        )),
                    },
                    TokenType::Minus => Ok(Value::Number(
                        self.require_number(left, op, "left operand")?
                            - self.require_number(right, op, "right operand")?,
                    )),
                    TokenType::Star => Ok(Value::Number(
                        self.require_number(left, op, "left operand")?
                            * self.require_number(right, op, "right operand")?,
                    )),
                    TokenType::Slash => {
                        let divisor = self.require_number(right, op, "right operand")?;
                        if divisor == 0.0 {
                            return Err(Self::runtime_error(op, "Division by zero."));
                        }
                        Ok(Value::Number(
                            self.require_number(left, op, "left operand")? / divisor,
                        ))
                    }
                    TokenType::Percent => {
                        let divisor = self.require_number(right, op, "right operand")?;
                        if divisor == 0.0 {
                            return Err(Self::runtime_error(op, "Modulo by zero."));
                        }
                        Ok(Value::Number(
                            self.require_number(left, op, "left operand")? % divisor,
                        ))
                    }
                    TokenType::Greater => Ok(Value::Bool(
                        self.require_number(left, op, "left operand")?
                            > self.require_number(right, op, "right operand")?,
                    )),
                    TokenType::GreaterEqual => Ok(Value::Bool(
                        self.require_number(left, op, "left operand")?
                            >= self.require_number(right, op, "right operand")?,
                    )),
                    TokenType::Less => Ok(Value::Bool(
                        self.require_number(left, op, "left operand")?
                            < self.require_number(right, op, "right operand")?,
                    )),
                    TokenType::LessEqual => Ok(Value::Bool(
                        self.require_number(left, op, "left operand")?
                            <= self.require_number(right, op, "right operand")?,
                    )),
                    TokenType::EqualEqual => Ok(Value::Bool(left == right)),
                    TokenType::BangEqual => Ok(Value::Bool(left != right)),
                    _ => Err(Self::runtime_error(op, "Unsupported binary operator.")),
                }
            }
            Expr::Call { callee, args } => {
                self.call_function(&callee.lexeme, args)
            }
            Expr::Grouping(inner) => self.evaluate(inner),
            Expr::Literal(value) => Ok(value.clone()),
            Expr::Unary { op, right } => {
                let right = self.evaluate(right)?;
                match op.kind {
                    TokenType::Bang => Ok(Value::Bool(!right.is_truthy())),
                    TokenType::Minus => {
                        Ok(Value::Number(-self.require_number(right, op, "operand")?))
                    }
                    _ => Err(Self::runtime_error(op, "Unsupported unary operator.")),
                }
            }
            Expr::Variable(name) => self.environment.borrow().get(name),
        }
    }

    fn call_function(&mut self, name: &str, args: &[Expr]) -> Result<Value> {
        // Evaluate arguments
        let mut evaluated_args = Vec::new();
        for arg in args {
            evaluated_args.push(self.evaluate(arg)?);
        }

        // Call built-in graphics functions
        match name {
            "create_window" => {
                if evaluated_args.len() != 2 {
                    return Err(DiagnosticError::new(
                        ErrorPhase::Runtime,
                        0,
                        0,
                        format!("create_window expects 2 arguments, got {}", evaluated_args.len()),
                    ));
                }
                let width = self.require_number_from_value(&evaluated_args[0])?;
                let height = self.require_number_from_value(&evaluated_args[1])?;
                idot_graphics::create_window(width as u32, height as u32);
                Ok(Value::Nil)
            }
            "draw_rect" => {
                if evaluated_args.len() != 5 {
                    return Err(DiagnosticError::new(
                        ErrorPhase::Runtime,
                        0,
                        0,
                        format!("draw_rect expects 5 arguments, got {}", evaluated_args.len()),
                    ));
                }
                let x = self.require_number_from_value(&evaluated_args[0])?;
                let y = self.require_number_from_value(&evaluated_args[1])?;
                let width = self.require_number_from_value(&evaluated_args[2])?;
                let height = self.require_number_from_value(&evaluated_args[3])?;
                let color = self.require_string_from_value(&evaluated_args[4])?;
                idot_graphics::draw_rect(x as f32, y as f32, width as f32, height as f32, &color);
                Ok(Value::Nil)
            }
            "draw_circle" => {
                if evaluated_args.len() != 4 {
                    return Err(DiagnosticError::new(
                        ErrorPhase::Runtime,
                        0,
                        0,
                        format!("draw_circle expects 4 arguments, got {}", evaluated_args.len()),
                    ));
                }
                let x = self.require_number_from_value(&evaluated_args[0])?;
                let y = self.require_number_from_value(&evaluated_args[1])?;
                let radius = self.require_number_from_value(&evaluated_args[2])?;
                let color = self.require_string_from_value(&evaluated_args[3])?;
                idot_graphics::draw_circle(x as f32, y as f32, radius as f32, &color);
                Ok(Value::Nil)
            }
            "draw_line" => {
                if evaluated_args.len() != 5 {
                    return Err(DiagnosticError::new(
                        ErrorPhase::Runtime,
                        0,
                        0,
                        format!("draw_line expects 5 arguments, got {}", evaluated_args.len()),
                    ));
                }
                let x1 = self.require_number_from_value(&evaluated_args[0])?;
                let y1 = self.require_number_from_value(&evaluated_args[1])?;
                let x2 = self.require_number_from_value(&evaluated_args[2])?;
                let y2 = self.require_number_from_value(&evaluated_args[3])?;
                let color = self.require_string_from_value(&evaluated_args[4])?;
                idot_graphics::draw_line(x1 as f32, y1 as f32, x2 as f32, y2 as f32, &color);
                Ok(Value::Nil)
            }
            "clear_window" => {
                if evaluated_args.len() != 1 {
                    return Err(DiagnosticError::new(
                        ErrorPhase::Runtime,
                        0,
                        0,
                        format!("clear_window expects 1 argument, got {}", evaluated_args.len()),
                    ));
                }
                let color = self.require_string_from_value(&evaluated_args[0])?;
                idot_graphics::clear_window(&color);
                Ok(Value::Nil)
            }
            _ => Err(DiagnosticError::new(
                ErrorPhase::Runtime,
                0,
                0,
                format!("Unknown function: '{}'", name),
            )),
        }
    }

    fn require_number_from_value(&self, value: &Value) -> Result<f64> {
        if let Value::Number(number) = value {
            return Ok(*number);
        }
        Err(DiagnosticError::new(
            ErrorPhase::Runtime,
            0,
            0,
            "Expected number".to_string(),
        ))
    }

    fn require_string_from_value(&self, value: &Value) -> Result<String> {
        if let Value::String(s) = value {
            return Ok(s.clone());
        }
        Err(DiagnosticError::new(
            ErrorPhase::Runtime,
            0,
            0,
            "Expected string".to_string(),
        ))
    }

    fn require_number(&self, value: Value, token: &Token, context: &str) -> Result<f64> {
        if let Value::Number(number) = value {
            return Ok(number);
        }
        Err(Self::runtime_error(
            token,
            format!("Expected number for {context}."),
        ))
    }

    fn runtime_error(token: &Token, message: impl Into<String>) -> DiagnosticError {
        DiagnosticError::new(ErrorPhase::Runtime, token.line, token.column, message)
    }

    fn io_error(prefix: &str, error: std::io::Error) -> DiagnosticError {
        DiagnosticError::new(ErrorPhase::Runtime, 0, 0, format!("{prefix}: {error}"))
    }
}

impl Default for Interpreter {
    fn default() -> Self {
        Self::new()
    }
}
